ebos/collecttoiorank.cc

```diff
@@ -805,6 +805,128 @@ class PackUnpackInterRegFlows : public P2PCommunicatorType::DataHandleInterface
     { this->globalInterRegFlows_.read(buffer); }
 };
 
+class PackUnpackFlowsn : public P2PCommunicatorType::DataHandleInterface
+{
+    const std::array<std::pair<std::string, std::pair<std::vector<int>, std::vector<double>>>, 3>& localFlowsn_;
+    std::array<std::pair<std::string, std::pair<std::vector<int>, std::vector<double>>>, 3>& globalFlowsn_;
+
+public:
+    PackUnpackFlowsn(const std::array<std::pair<std::string, std::pair<std::vector<int>, std::vector<double>>>, 3> & localFlowsn,
+                     std::array<std::pair<std::string, std::pair<std::vector<int>, std::vector<double>>>, 3>& globalFlowsn,
+                     const bool isIORank)
+        : localFlowsn_(localFlowsn)
+        , globalFlowsn_(globalFlowsn)
+    {
+        if (! isIORank) { return; }
+
+        MessageBufferType buffer;
+        this->pack(0, buffer);
+
+        // pass a dummy_link to satisfy virtual class
+        const int dummyLink = -1;
+        this->unpack(dummyLink, buffer);
+    }
+
+    void pack(int link, MessageBufferType& buffer)
+    {
+        if (link != 0)
+            throw std::logic_error("link in method pack is not 0 as expected");
+        for (int i = 0; i < 3; ++i) {
+            const auto& name = localFlowsn_[i].first;
+            buffer.write(name);
+            unsigned int size = localFlowsn_[i].second.first.size();
+            buffer.write(size);
+            for (unsigned int j = 0; j < size; ++j) {
+                const auto& nncIdx = localFlowsn_[i].second.first[j];
+                buffer.write(nncIdx);
+                const auto& flows = localFlowsn_[i].second.second[j];
+                buffer.write(flows);
+            }
+        }
+    }
+
+    void unpack(int /*link*/, MessageBufferType& buffer)
+    {
+        for (int i = 0; i < 3; ++i) {
+            std::string name;
+            buffer.read(name);
+            globalFlowsn_[i].first = name;
+            unsigned int size = 0;
+            buffer.read(size);
+            for (unsigned int j = 0; j < size; ++j) {
+                int nncIdx;
+                double data;
+                buffer.read(nncIdx);
+                buffer.read(data);
+                if (nncIdx < 0)
+                    continue;
+                globalFlowsn_[i].second.second[nncIdx] = data;
```

Comment:
Is there a guarantee that the array element
```
globalFlowsn_[i].second.second[nncIdx]
```
exists, i.e., has been properly allocated, on the receiving process?  If so, I think it'd be appropriate to add a comment here stating what that guarantee is.